Overhaul `insert.rst`

norosa opened the pull request:
Specifically:

- Correct use of RST labels

- Restructure to make better sense

  - *Parameters* now appears under *Synopsis*

  - The caution about dynamic selects has been moved to an admonition further up the page (where it is more relevant)

- Copyedits for text to improve correctness, clarity, and style (e.g., referring to things as "statements", switches to active voice, resolving indeterminate references)

  - Fix ambiguous use of ``WITH`` in ``ON CONFLICT DO UPDATE SET`` section (closes https://github.com/crate/crate/issues/11387)

- Improve use of SQL examples

  - Consistent use of uppercase for SQL reserved words
  
  - Better (more descriptive) use of table names, column names, and so on

norosa commented:
@mfussenegger can you verify this PR fixes https://github.com/crate/crate/issues/11387, please?

mfussenegger commented on docs/sql/statements/insert.rst:
```diff
@@ -26,130 +31,155 @@ Synopsis
 
 If the :ref:`expression <gloss-expression>` for any column is not of the
-correct data type, automatic type conversion will be attempted.
+correct data type, CrateDB will not attempt automatic :ref:`type conversion
```
```suggestion
correct data type, CrateDB will attempt automatic :ref:`type conversion
```

mfussenegger commented on docs/sql/statements/insert.rst:
```diff
@@ -26,130 +31,155 @@ Synopsis
 
-:output_name:
-    A name to use for the result of the output expression.
+You may limit the scope of this clause by specifying one or more columns after
+``ON CONFLICT``, like so::
+
+     INSERT INTO my_table (col_a, col_a) VALUES (1, 42)
+     ON CONFLICT (col_a) DO NOTHING
+
+In this example, CrateDB would only silently ignore conflicting inserts for
+``col_a``.
```
Only primary key columns can be used in the `ON CONFLICT` clause - this section here sounds as if the columns are arbitrary.

norosa replied:
ooh good catch thx

mfussenegger commented on docs/sql/statements/insert.rst:
```diff
@@ -26,130 +31,155 @@ Synopsis
 
-:output_name:
-    A name to use for the result of the output expression.
+You may limit the scope of this clause by specifying one or more of the primary
+key columns after ``ON CONFLICT``, like so::
+
+     INSERT INTO my_table (col_a, col_a) VALUES (1, 42)
+     ON CONFLICT (col_a) DO NOTHING
+
+In this example, CrateDB would only silently ignore conflicting inserts for
+``col_a``. CrateDB would not ignore conflicting values for other primary keys
+(if they exist).
```
Did you test if this actually works? I'm not sure if we support this.

(Please push a dedicated fixup commit next time instead of squashing,  - makes it easier to review the follow up changes)

mfussenegger replied:
@norosa could you get back to this so we can wrap this up?

norosa commented:
I pushed a fixup commit. I am copying the commit message here so it is easier to read and so that it will be easy to find after I eventually rebase to squash

---

Details:

- Added RST labels to `compatibility.rst` so that I could add a cross-reference to this document.

- Added RST labels to `dml.rst` for the same reason.

  While I was in that file, I corrected the SQL examples to use upper-case for reserved keywords.

  I also added a cross-reference back to `insert.rst` syntax doc.

- Reworked my changes to `insert.rst`.

  As suggested, I took the time to test my understanding of this feature through direct experimentation.

  The ability to specify columns in the `ON CONFLICT` clause implied (to me) that there is some choice available, which is incorrect. With CrateDB, the`ON CONFLICT` columns must always exactly match the primary key.

  This apparent redundancy confused me, so I conducted the same experiment with PostgreSQL to better understand how PostgreSQL implements `ON CONFLICT` and how this feature is used.

  Eventually, I realized that PostgreSQL allows you to specify columns in the `ON CONFLICT` clause because PostgreSQL supports a range of different constraints that CrateDB does not support.

  Because CrateDB does not support these constraints, the ability to specify columns in the `ON CONFLICT` clause *is* redundant but also necessary for syntax compatibility.

  The only mention of this I could find in the existing docs was a brief list in `compatibility.rst`.

  To address this issue, I have added a `NOTE` admonition to `insert.rst` which makes it explicit that CrateDB only supports `ON CONFLICT` with primary key constraints and that the `conflict_target` must always exactly match the primary key.

  Additionally, I realized the existing explanation of the `ON CONFLICT` clause does a poor job of explaining its purpose and operation. To address this, I made the following changes:

  - Reworded the intro to the `ON CONFLICT` section for clarity

  - Adjusted the first example so that it is a syntax example *only*. Previously, it mixed example SQL with syntax grammar.

  - Added an example query that demonstrates how to use `ON CONFLICT` to perform an operation on the existing record. In this case, incrementing `visits` by one.

    I took this example query from the existing upserts documentation in `dml.rst`. However, I found the query in `dml.rst` confusing because it specifies two values for `last_visit` (one in the `INSERT` and one in the `DO UPDATE`), both different. This makes no sense in practice.

    I fixed this by modifying the query in both places so that the first example only increments `visits` (demonstrating the ability to reference the existing value).

    I then modified the second example in `insert.rst` so that it demonstrates the proper use of `excluded` to overwrite `last_visits` (mirroring the examples in the upserts doc).

  - I added an explicit description of what each query does to better guide the reader.

  - Added a cross-reference to the upserts doc.

  - Rewrote `ON CONFLICT DO NOTHING` to correct my previous mistake.

    I added a `NOTE` admonition to explain that the conflict target is optional, but must match the primary key if specified.

mfussenegger reviewed (approved):
:+1:

norosa commented:
squashed and updated commit message to reflect sum changes
